### PR TITLE
JSON Sidebar regression fixes

### LINKED ISSUE #58598: Sidebar UI, W_NO_PADDING flag has issues
### Describe the bug

W_NO_PADDING flag breaks "arrange": "rows" in child widgets

### Steps to reproduce

Use attached sidebar.

### Expected behavior

Flag should only affect "arrange": "columns" widgets

### Screenshots

image

image

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19044.1766 (21H2)
- Game Version: 80cf483 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions]
]

### Additional context

[!_sidebar.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/8949537/_sidebar.zip)

related to the PR #58543

### PR DESCRIPTION
**Pull request by Inglonias**
#### Summary
Interface "Fixes regressions introduced by swapping to JSON sidebars"

#### Purpose of change
The major regressions found so far are in the health widget, where limb labels would not change color based on status, and the compass, whose labels also wouldn't change color anymore.

#### Describe the solution
New text widgets were created for each limb that change color based on their status, replicating the old sidebar's behavior. Similar widgets were created for the compass. These widgets rely on new dialogue conditions called `u_monsters_in_direction` and `u_safe_mode_trigger`

#### Describe alternatives you've considered
Hard coded special cases for the labels of the health and compass widget. I also considered inventing a new type of widget to make the JSON less tedious for this project, but since I'm only writing the JSON once, I figured it wasn't worth it. Yet.

#### Testing
The health and compass widgets seem to work.

#### Additional context
This also fixes #58598 since I needed to do that to get this working. If you want me to split that out into its own PR, let me know.

### CODE REVIEW

**Comment by Inglonias** [comment]
There are two known regressions remaining that are not currently possible to fix:

* Cannot view mending status of broken limbs
* * This one is pretty self-explanatory. We can't fix this because right now there is no way to completely swap widgets out on the fly. I can't show one graph if the limb is not broken, and another when it is. This would need to be added to fix this. In the meantime, this status can be viewed in the Medical UI (default `Shift+N` keybinding)
* Weird compass label functionality
* * Compass labels used to turn red to indicate which direction the monster that triggered safe mode last turn was in. Without dialogue conditions relating to safe mode rules (please don't make me), this is not possible.

**Comment by snipercup** [comment]
I didn't mind your sidebar pr getting merged since you said the game will have the same functionality as before. Now you're saying some functionality will be lost unless the difficult solution is implemented. Can you please implement everything needed  to make it work as before? I am holding off on updating because the compass functionality is a major part of my gameplay style, i would almost call it essential.

**Comment by Inglonias** [comment]
> I didn't mind your sidebar pr getting merged since you said the game will have the same functionality as before. Now you're saying some functionality will be lost unless the difficult solution is implemented. Can you please implement everything needed to make it work as before? I am holding off on updating because the compass functionality is a major part of my gameplay style, i would almost call it essential.

I just fixed the compass regression. The bone mending regression is much trickier, since widgets were never expected to need to show more than one var at a time.

**Comment by dseguin** [comment]
Just in case you'd like to compare the number of monsters in a direction instead of just true-or-false, you can move the condition within `get_get_int`:

<details>
<summary>(diff, click to expand)</summary>

```diff
diff --git a/data/json/ui/compass_danger.json b/data/json/ui/compass_danger.json
index b25d5b822e..aa1e31378b 100644
--- a/data/json/ui/compass_danger.json
+++ b/data/json/ui/compass_danger.json
@@ -5,7 +5,7 @@
     "style": "text",
     "direction": "N",
     "default_clause": { "text": "N:", "color": "c_dark_gray" },
-    "clauses": [ { "id": "danger", "text": "N:", "color": "c_light_gray", "condition": { "u_monsters_in_direction": "N" } } ]
+    "clauses": [ { "id": "danger", "text": "N:", "color": "c_light_gray", "condition": { "compare_int": [ { "monsters_in_direction": "N" }, ">", { "const": 0 } ] } } ]
   },
   {
     "id": "compass_S_danger_label",
@@ -13,7 +13,7 @@
     "style": "text",
     "direction": "S",
     "default_clause": { "text": "S:", "color": "c_dark_gray" },
-    "clauses": [ { "id": "danger", "text": "S:", "color": "c_light_gray", "condition": { "u_monsters_in_direction": "S" } } ]
+    "clauses": [ { "id": "danger", "text": "S:", "color": "c_light_gray", "condition": { "compare_int": [ { "monsters_in_direction": "S" }, ">", { "const": 0 } ] } } ]
   },
   {
     "id": "compass_E_danger_label",
@@ -21,7 +21,7 @@
     "style": "text",
     "direction": "E",
     "default_clause": { "text": "E:", "color": "c_dark_gray" },
-    "clauses": [ { "id": "danger", "text": "E:", "color": "c_light_gray", "condition": { "u_monsters_in_direction": "E" } } ]
+    "clauses": [ { "id": "danger", "text": "E:", "color": "c_light_gray", "condition": { "compare_int": [ { "monsters_in_direction": "E" }, ">", { "const": 0 } ] } } ]
   },
   {
     "id": "compass_W_danger_label",
@@ -29,7 +29,7 @@
     "style": "text",
     "direction": "W",
     "default_clause": { "text": "W:", "color": "c_dark_gray" },
-    "clauses": [ { "id": "danger", "text": "W:", "color": "c_light_gray", "condition": { "u_monsters_in_direction": "W" } } ]
+    "clauses": [ { "id": "danger", "text": "W:", "color": "c_light_gray", "condition": { "compare_int": [ { "monsters_in_direction": "W" }, ">", { "const": 0 } ] } } ]
   },
   {
     "id": "compass_NE_danger_label",
@@ -37,7 +37,7 @@
     "style": "text",
     "direction": "NE",
     "default_clause": { "text": "NE:", "color": "c_dark_gray" },
-    "clauses": [ { "id": "danger", "text": "NE:", "color": "c_light_gray", "condition": { "u_monsters_in_direction": "NE" } } ]
+    "clauses": [ { "id": "danger", "text": "NE:", "color": "c_light_gray", "condition": { "compare_int": [ { "monsters_in_direction": "NE" }, ">", { "const": 0 } ] } } ]
   },
   {
     "id": "compass_NW_danger_label",
@@ -45,7 +45,7 @@
     "style": "text",
     "direction": "NW",
     "default_clause": { "text": "NW:", "color": "c_dark_gray" },
-    "clauses": [ { "id": "danger", "text": "NW:", "color": "c_light_gray", "condition": { "u_monsters_in_direction": "NW" } } ]
+    "clauses": [ { "id": "danger", "text": "NW:", "color": "c_light_gray", "condition": { "compare_int": [ { "monsters_in_direction": "NW" }, ">", { "const": 0 } ] } } ]
   },
   {
     "id": "compass_SE_danger_label",
@@ -53,7 +53,7 @@
     "style": "text",
     "direction": "SE",
     "default_clause": { "text": "SE:", "color": "c_dark_gray" },
-    "clauses": [ { "id": "danger", "text": "SE:", "color": "c_light_gray", "condition": { "u_monsters_in_direction": "SE" } } ]
+    "clauses": [ { "id": "danger", "text": "SE:", "color": "c_light_gray", "condition": { "compare_int": [ { "monsters_in_direction": "SE" }, ">", { "const": 0 } ] } } ]
   },
   {
     "id": "compass_SW_danger_label",
@@ -61,7 +61,7 @@
     "style": "text",
     "direction": "SW",
     "default_clause": { "text": "SW:", "color": "c_dark_gray" },
-    "clauses": [ { "id": "danger", "text": "SW:", "color": "c_light_gray", "condition": { "u_monsters_in_direction": "SW" } } ]
+    "clauses": [ { "id": "danger", "text": "SW:", "color": "c_light_gray", "condition": { "compare_int": [ { "monsters_in_direction": "SW" }, ">", { "const": 0 } ] } } ]
   },
   {
     "id": "compass_N_danger_layout",
diff --git a/src/condition.cpp b/src/condition.cpp
index 863825199b..323a0cb2e9 100644
--- a/src/condition.cpp
+++ b/src/condition.cpp
@@ -392,18 +392,6 @@ void conditional_t<T>::set_u_has_mission( const JsonObject &jo )
     };
 }
 
-template<class T>
-void conditional_t<T>::set_u_monsters_in_direction( const JsonObject &jo )
-{
-    const std::string &dir = jo.get_string( "u_monsters_in_direction" );
-    condition = [dir]( const T & ) {
-        //This string_to_enum function is defined in widget.h. Should it be moved?
-        const int card_dir = static_cast<int>( io::string_to_enum<cardinal_direction>( dir ) );
-        int monster_count = get_avatar().get_mon_visible().unique_mons[card_dir].size();
-        return monster_count > 0;
-    };
-}
-
 template<class T>
 void conditional_t<T>::set_has_strength( const JsonObject &jo, const std::string &member,
         bool is_npc )
@@ -1359,6 +1347,12 @@ std::function<int( const T & )> conditional_t<T>::get_get_int( const JsonObject
                 return static_cast<int>( get_weather().weather_precise->pressure );
             };
         }
+    } else if( jo.has_member( "monsters_in_direction" ) ) {
+        std::string dir = jo.get_string( "monsters_in_direction" );
+        int card_dir = static_cast<int>( io::string_to_enum<cardinal_direction>( dir ) );
+        return [card_dir]( const T & ) {
+            return static_cast<int>( get_avatar().get_mon_visible().unique_mons[card_dir].size() );
+        };
     } else if( jo.has_member( "u_val" ) || jo.has_member( "npc_val" ) ||
                jo.has_member( "global_val" ) ) {
         const bool is_npc = jo.has_member( "npc_val" );
@@ -2506,8 +2500,6 @@ conditional_t<T>::conditional_t( const JsonObject &jo )
         set_is_riding( is_npc );
     } else if( jo.has_string( "u_has_mission" ) ) {
         set_u_has_mission( jo );
-    } else if( jo.has_string( "u_monsters_in_direction" ) ) {
-        set_u_monsters_in_direction( jo );
     } else if( jo.has_int( "u_has_strength" ) || jo.has_object( "u_has_strength" ) ) {
         set_has_strength( jo, "u_has_strength" );
     } else if( jo.has_int( "npc_has_strength" ) || jo.has_object( "npc_has_strength" ) ) {
diff --git a/src/condition.h b/src/condition.h
index 1b31a163b1..adc7f54a03 100644
--- a/src/condition.h
+++ b/src/condition.h
@@ -30,7 +30,7 @@ const std::unordered_set<std::string> simple_string_conds = { {
 };
 const std::unordered_set<std::string> complex_conds = { {
         "u_has_any_trait", "npc_has_any_trait", "u_has_trait", "npc_has_trait",
-        "u_has_flag", "npc_has_flag", "npc_has_class", "u_has_mission", "u_monsters_in_direction",
+        "u_has_flag", "npc_has_flag", "npc_has_class", "u_has_mission",
         "u_has_strength", "npc_has_strength", "u_has_dexterity", "npc_has_dexterity",
         "u_has_intelligence", "npc_has_intelligence", "u_has_perception", "npc_has_perception",
         "u_is_wearing", "npc_is_wearing", "u_has_item", "npc_has_item", "u_has_move_mode", "npc_has_move_mode",
@@ -258,7 +258,6 @@ struct conditional_t {
         void set_is_riding( bool is_npc = false );
         void set_npc_has_class( const JsonObject &jo, bool is_npc );
         void set_u_has_mission( const JsonObject &jo );
-        void set_u_monsters_in_direction( const JsonObject &jo );
         void set_has_strength( const JsonObject &jo, const std::string &member, bool is_npc = false );
         void set_has_dexterity( const JsonObject &jo, const std::string &member, bool is_npc = false );
         void set_has_intelligence( const JsonObject &jo, const std::string &member, bool is_npc = false );
```

</details>

That will let you compare the number with
```JSON
"condition": {
  "compare_int": [ { "monsters_in_direction": "N" }, ">", { "const": 0 } ]
}
```

**Comment by Inglonias** [comment]
@snipercup Now that it's actually fixed, I'm curious as to why you find that feature so useful? I admit it's handy to have, but I would hardly call it essential myself. Typically my first instinct after seeing a safemode trigger is to open the UI that lists all nearby monsters.

File: src/widget.cpp [hunk 1]
```diff
@@ -1493,7 +1493,7 @@ std::string widget::layout( const avatar &ava, unsigned int max_width, int label
                 const std::string txt = cur_child.layout( ava, skip_pad_this ? 0 : cur_width,
                                         label_width, skip_pad_this );
                 // Store the resulting text for this column
-                cols.emplace_back( foldstring( txt, skip_pad_this ? 0 : cur_width ) );
+                cols.emplace_back( string_split( txt, '\n' ) );
```
Comment (Inglonias) [inline src/widget.cpp:1496]: It looks like this change is causing some test failures.

**Comment by snipercup** [comment]
> @snipercup Now that it's actually fixed, I'm curious as to why you find that feature so useful? I admit it's handy to have, but I would hardly call it essential myself. Typically my first instinct after seeing a safemode trigger is to open the UI that lists all nearby monsters.

You're right. I should have explained first. Basically i just walk around until it lights up and then i stop what i am doing and hold 5 until they are close to optimize my combat strategy. Without it i would not know if a zombie has seen me intil it is too late or i would have to press V every step. Thank you for fixing it 👍

File: src/widget.cpp (hunk 1 above)
Comment (Inglonias) [reply src/widget.cpp:1496]: I changed the tests to fit the fix. The failure is in a layout that does some strange things. This is worth taking a look at before merging.

**Comment by Inglonias** [comment]
I missed the compact sidebar in this PR. My bad. Working on that now.

**Pull request merged**